main.py

```diff
@@ -2040,9 +2062,10 @@ def show_hover(self, point, contents):
 def preserve_whitespace(contents: str) -> str:
     """Preserve empty lines and whitespace for markdown conversion."""
     contents = contents.strip(' \t\r\n')
-    contents = contents.replace('\t', '&nbsp;' * 4)
-    contents = contents.replace('  ', '&nbsp;' * 2)
-    contents = contents.replace('\n\n', '\n&nbsp;\n')
+    contents = contents.replace('\r\n', '\n')
+    contents = contents.replace('\t', '\u00A0' * 4)
+    contents = contents.replace('  ', '\u00A0' * 2)
+    contents = contents.replace('\n\n', '\n\u00A0\n')
```

Comment:
I am no longer sure about that line. I originally added it as I experienced absolutely no spaces between paragraphs. Indeed it breaks the creation of `<p>paragraph1</p><p>paragrah2</p>` The line causes markdown to create `<p>paragraph1<br><br>paragrah2</p>`. 

If we remove this line and tweak the CSS a little bit to add proper padding, we may not even need the `md2html()` functions as merging the lines into `div`s is effectively nothing else then creating several paragraphs. The solution would be far easier then anyone of us thought yesterday 😄  

You may try ...

```python
            ...
            if language:
                formatted.append("```{}\n{}\n```\n".format(language, value))
            else:
                formatted.append(preserve_whitespace(value))

        mdpopups.show_popup(
            self.view,
            "".join(formatted),
            css='''
                .lsp_hover {
                    margin: 0.5rem;
                }
                .lsp_hover .highlight {
                    border-width: 0;
                    margin-bottom: 0;
                }
                .mdpopups .lsp_hover p {
                    padding: 0.5rem 0.5rem 0 0.5rem;
                }
            ''',
            md=True,
            flags=sublime.HIDE_ON_MOUSE_MOVE_AWAY,
            location=point,
            wrapper_class="lsp_hover",
            max_width=800)


def preserve_whitespace(contents: str) -> str:
    """Preserve empty lines and whitespace for markdown conversion."""
    contents = contents.strip(' \t\r\n')
    contents = contents.replace('\r\n', '\n')
    contents = contents.replace('\t', '\u00A0' * 4)
    contents = contents.replace('  ', '\u00A0' * 2)
    return contents
```

Reply:
That's also pretty clean. However,  it looks a lot better if the description text aligns with the code text. Now it aligns with the code box:

<img width="129" alt="screen shot 2017-10-10 at 20 00 37" src="https://user-images.githubusercontent.com/2543659/31402232-b6e751b0-adf5-11e7-9251-c46719f13224.png">

Some screenshots to compare:

<img width="801" alt="screen shot 2017-10-10 at 20 04 27" src="https://user-images.githubusercontent.com/2543659/31402639-e7bb5ccc-adf6-11e7-8758-7888667b0c48.png">

<img width="793" alt="screen shot 2017-10-10 at 20 11 49" src="https://user-images.githubusercontent.com/2543659/31402770-46ec5f7a-adf7-11e7-9299-a20b25cf2380.png">

<img width="817" alt="screen shot 2017-10-10 at 20 11 20" src="https://user-images.githubusercontent.com/2543659/31402780-4cd9a186-adf7-11e7-927c-3cb0960ce731.png">

Also note in the first screenshot how there is whitespace missing after `function _(exported, ambient)_`. But disregarding that, I'm leaning towards the second or third style. 


----

As an aside, I don't think we have to prefix css selectors with `.mdpopups` (e.g. `.mdpopups .lsp_hover`). Our css follows after mdpopup's so there is no need to increase the specificity.

Reply:
As for sticking closely to mdpopups, this is in line with @deathaxe's suggestion, but with some fixes and even less lines of code:

```python
        for item in contents:
            value = ""
            language = None
            if isinstance(item, str):
                value = item
            else:
                value = item.get("value")
                language = item.get("language")

            if language:
                formatted.append("```{}\n{}\n```\n".format(language, value))
            else:
                formatted.append(preserve_whitespace(value))

        mdpopups.show_popup(
            self.view,
            "\n".join(formatted),
            css='''
                .lsp_hover {
                    margin: 0.5rem 0.5rem 0 0.5rem;
                }
                .lsp_hover p {
                    margin-bottom: 0.5rem;
                }
            ''',
            md=True,
            flags=sublime.HIDE_ON_MOUSE_MOVE_AWAY,
            location=point,
            wrapper_class="lsp_hover",
            max_width=800)


def preserve_whitespace(contents: str) -> str:
    """Preserve empty lines and whitespace for markdown conversion."""
    contents = contents.strip(' \t\r\n')
    contents = contents.replace('\r\n', '\n')
    contents = contents.replace('\t', '\u00A0' * 4)
    contents = contents.replace('  ', '\u00A0' * 2)
    return contents
```

<img width="819" alt="screen shot 2017-10-10 at 20 38 35" src="https://user-images.githubusercontent.com/2543659/31403960-0435a34a-adfb-11e7-8ef5-71e308e5c917.png">

<img width="573" alt="screen shot 2017-10-10 at 20 39 00" src="https://user-images.githubusercontent.com/2543659/31403984-12acc8c2-adfb-11e7-82fb-a3ab1f5c8897.png">

I still think it's prettier to align the text, but I also very much like how little code this is.

To align text, it's now enough to add `padding: 0 0.5rem;` to `.lsp_hover p`. 

<img width="820" alt="screen shot 2017-10-10 at 20 41 18" src="https://user-images.githubusercontent.com/2543659/31404234-c3b63004-adfb-11e7-873a-44ce0a07be2a.png">

I also kinda like removing the border around the code, by adding `.lsp_hover .highlight { border-width: 0; border-radius: 0; }`

<img width="807" alt="screen shot 2017-10-10 at 20 43 50" src="https://user-images.githubusercontent.com/2543659/31404279-d9d3b6d6-adfb-11e7-8943-8622ba75bd67.png">

Reply:
LGTM, you probably want to experiment around with the new signature help browser now, too.

Reply:
> I don't think we have to prefix css selectors with .mdpopups

agree.

> However, it looks a lot better if the description text aligns with the code text.

Can be tweaked by setting the margin of `<p>`, i think. The highlight is a `<div>` and is therefore not effected. I just didn't see as pyls does return the signature as ordinary text unfortunately.

> there is whitespace missing after 

I saw such issues with signatures, too. Would be interesting how the raw text looks like, and if there is a way to fix it before passing it to mdpopups.

Reply:
Changed my mind about the border around the code box. I'll push an update and will look into some of the other popups later on. Gotta watch a football match right now, will get back to the code as soon as we're a couple of goals behind ;)

Reply:
I see, it might be caused by applying `preserve_whitespace()` to each line separately. It strips tailing newlines, and therefore should be called with the joined lines.